### ISSUE TITLE
[mingw64/msys] git 2.16 updated package broke git fetch for https urls

### ISSUE THREAD

**Issue by gena01** [git-for-windows/git#1453]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

``` 
git --version
git version 2.16.1.windows.1
```
 I am using mingw-w64-x86_64-git 2.16.1.1.1a4ee4d5d8-1

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Running "MSYS2 installed "MINGW64" bash environment. Mapping to the official git-for-windows repo as stated in the wiki. 

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git fetch --all
```
 - What did you expect to occur after running these commands?

I expected it to work as before (it did work in 2.15 version/package)

 - What actually happened instead?

```
Fetching origin
fatal: unable to access 'https://github/xxxxxx': error setting certificate verify locations:
  CAfile: /mingw64/ssl/certs/ca-bundle.crt
  CApath: none
```

**Comment by PhilipOakley** [comment]
@gena01 gennaddy, this may also be a cURL/ssl/schannel certificates issue as there have also been a number of changes (and issues) there as well. Have you checked if any of them apply? The reposrt suggests it is a local probelm in finding the certificate verify locations.

**Comment by gena01** [comment]
I don't see my environment as anything custom. I run msys2 pointing to whatever default repo(s) installer sets up + "git for windows" provided packages.

**Comment by PhilipOakley** [comment]
Ah, so this isn't 'Git for Windows' complete (This project), but you want simply 'Git' on MSYS2? would that be right?

**Comment by gena01** [comment]
@PhilipOakley I guess so. I wasn't sure if this was the right place to report it. The binaries I am using are definitely from this github organization. This is one of the commits that's not working for me: https://github.com/git-for-windows/MINGW-packages/commit/c25e06306c4b07bb77acb541348d5a72963a0ff3

**Comment by swaldman3** [comment]
I've landed here with the same problem; I had thought it was an issue with my corporate proxy configuration, but think it might be more general.

Setup: Git for Windows 2.16.0.2-64-bit installer. Installation options to use the non-cmd interpreter (I forget what it's called) and to use OpenSSL.

No matter what I set http.sslcainfo to, at repo, global or system level, trying to fetch from a remote always gives (with GIT_CURL_VERBOSE=1),
```
$ git fetch origin
* STATE: INIT => CONNECT handle 0x5667e0; line 1418 (connection #-5000)
* Couldn't find host github.com in the _netrc file; using defaults
* Added connection 0. The cache now contains 1 members
*   Trying 192.168.41.8...
* TCP_NODELAY set
* STATE: CONNECT => WAITCONNECT handle 0x5667e0; line 1470 (connection #0)
* Connected to 192.168.41.8 (192.168.41.8) port 80 (#0)
* STATE: WAITCONNECT => WAITPROXYCONNECT handle 0x5667e0; line 1580 (connection #0)
* Marked for [keep alive]: HTTP default
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* error setting certificate verify locations:
  CAfile: /mingw64/ssl/certs/ca-bundle.crt
  CApath: none
* Expire cleared
* multi_done
* Closing connection 0
* The cache now contains 0 members
fatal: unable to access 'https://github.com/swaldman3/MATLAB-scripts.git/': error setting certificate verify locations:
  CAfile: /mingw64/ssl/certs/ca-bundle.crt
  CApath: none
```

NB the path given there does not change, regardless of the http.sslCAInfo setting.
In desperation I tried setting http.sslVerify=false and nothing changes. `git config -l` does show the current settings, but it's as though those settings are being ignored by `git fetch` and the config is being picked up from somewhere else.

Apologies if this isn't too clear - I'm no expert here. Feel free to ask for further info if it'll help.

**Comment by dscho** [comment]
Git for Windows also ships custom cURL versions. Not sure whether this would fix your problem. Also, using `GIT_TRACE_CURL=1` might help diagnose it further.

You know, if at least one of you gentle people (who refuse to use Git for Windows as offered in our Releases, and prefers to use their own MSYS2-based setups) would step in and help, get involved, and maintained that part, we would really get somewhere. Get engaged!

**Comment by gena01** [comment]
@dscho I don't think this issue is specific to "MSYS2-based setup". Sounds like some part of the binaries that tie into MINGW64 could have had a regression between 2.15 and 2.16. 

2.16 has an updated openssl, so could be some patch missing or regression in the new openssl integration somehow.

**Comment by dscho** [comment]
> I don't think this issue is specific to "MSYS2-based setup"

This is, however, my working hypothesis until somebody shows me an MCVE *with Git for Windows*.

For the record, I fetch and push via HTTPS multiple times a day, and I did not see the mentioned error message *in a long time*. As a matter of fact, I do verify (manually) that a `git ls-remote https://...` works, every single time before I upload new Git for Windows versions. Therefore, it would be really surprising to me if this was a bug that affected Git for Windows and/or Git for Windows' SDK.

To protect my time, I will close this ticket until the time an MCVE is produced that shows that this affects Git for Windows, too.

**Comment by swaldman3** [comment]
I'm not sure what a MCVE is, but I reported it with Git for Windows above, 8 days ago.

**Comment by dscho** [comment]
> I'm not sure what a MCVE is

To help with that, we say this in the issue reporting template that was filled out in the very beginning:

> What commands did you run to trigger this issue? If you can provide a Minimal, Complete, and Verifiable example this will help us understand the issue.

To clarify even further, we link to StackOverflow's description: http://stackoverflow.com/help/mcve

> but I reported it with Git for Windows above, 8 days ago.

This does not reproduce here, and is therefore not verifiable.

So where does this leave you?

You could try to run [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) as I suggested [in another ticket](https://github.com/git-for-windows/git/issues/1363#issuecomment-353190073) (that looks very much like discussing the same thing).

You could also have a look at your `http.sslCAInfo` setting. Here, it looks like this:

```sh
$ git config --show-origin -l | grep ca-bund
file:"C:\\ProgramData/Git/config"       http.sslcainfo=c:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
file:"C:\\Program Files\\Git\\mingw64/etc/gitconfig"    http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
```

Two things to note:

1. The `C:\ProgramData/Git/config` setting is a left-over from a previous installation and does not matter
2. The `C:\Program Files\Git\mingw64/etc/gitconfig` setting is the relevant one, and most notably it contains *the full path* to `ca-bundle.crt`, not just the path relative to the pseudo-Unix-root (i.e. the path is *not* just `/mingw64/ssl/certs/ca-bundle.crt`)

Please see whether this `http.sslCAInfo` setting is obviously incorrect in your setup. If it is correct, we will have to get the big hammer out and instrument the source code both of Git and of cURL, and I will have to ask you to run a few tests, but I will only spend that effort if needed, and only if there is a good chance that I can get you to perform these tests...

**Comment by swaldman3** [comment]
I can't provide a MCVE, because I have a strong suspicion that to reproduce it you must first be behind a corporate proxy that is MITM-intercepting https traffic.
If you are in, or can simulate, that position, then to reproduce simply enter a git repo directory and run `git fetch origin`.

As noted in https://github.com/git-for-windows/git/issues/1453#issuecomment-360102323, I seemed to get exactly the same error *regardless* of what was in http.sslCAInfo - whether sane or not. It was as though something (git or curl?) was ignoring this setting.

For me the fix (for now) was to downgrade to release 2.15.1.2. I'm usually happy to spend my time helping OSS authors to test things - after all, I'm grateful for use of the software and so forth. But in this case, given the somewhat belligerent response I've received, I'm not inclined to break my now-working installation again. Sorry.

**Comment by dscho** [comment]
> As noted in #1453 (comment), I seemed to get exactly the same error regardless of what was in http.sslCAInfo - whether sane or not. It was as though something (git or curl?) was ignoring this setting.

What does `git config --show-origin http.sslCAInfo` say?

> I'm usually happy to spend my time helping OSS authors to test things - after all, I'm grateful for use of the software and so forth. But in this case, given the somewhat belligerent response I've received, I'm not inclined to break my now-working installation again.

Oh, come on. You certainly must understand how frustrating it is to try to help people, only to get as little information as you provided, and only what I asked for specifically.

Don't be such a dick.

Back to the topic. I can see how it would appear likely that a MITM proxy would cause problems, but given that 2.15.1.2 does not have those problems, I would reject that notion.

If you are really worried about breaking your setup, just use Secure Channel instead of OpenSSL and get unblocked right away, even with an up-to-date Git for Windows version.

For the record, you can always test things without breaking your now-working setup using the Portable Git. That use case is the main reason for Portable Git's existence.

If you want to show gratitude for the insane amount of work that went into Git and Git for Windows, step up your game and provide at least the following information:

- is there a `mingw64/libexec/git-core/libcurl-4.dll`?
- If so, does [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) show that it gets loaded?
- What does `sha256sum /mingw64/libexec/git-core/libcurl-4.dll` show (just making sure that you have the same version as I use here for testing)?
- What are the exact outputs of the `git config` commands (please do not provide hear-say, just copy-paste the output, otherwise I don't have enough information to work with).

**Comment by swaldman3** [comment]
> Don't be such a dick.

<snip>

> If you want to show gratitude for the insane amount of work that went into Git and Git for Windows, step > up your game and provide at least the following information:

If you want me to go out of my way to install a non-working version of the software to help you diagnose problems, don't call me a dick, or tell me to "step up my game".

Thank you for putting Git for Windows together.
Goodbye.

**Comment by dscho** [comment]
> Thank you for putting Git for Windows together.

I appreciate your thanking me. Actions speak louder than words, though.

Open Source is a deal: you get things without paying money, but of course those things are not worthless, so you have to pay in different ways.

But I see that this conversation went south the moment you claimed that you provided an MCVE in the same sentence where you said that you do not even know what an MCVE is. Sadly, this is wasting my time very freely (which I do not appreciate one bit), and I felt so insulted that I insulted you back. Granted, I really feel that your behavior was inappropriate, but that does not excuse *my* behavior, for which I hereby apologize.

As it is, this turned into a very unpleasant exchange, which not only is unproductive, but also poisons the project for whoever is more willing to fulfill their "Open Source contract". This poisoning is something which I cannot allow, it is my duty as maintainer to try my best (which is not very good, admittedly, but at least I try my best) to prevent.

Therefore I will lock this conversation, and if necessary block users.
